Holding this after rebasing onto current main. The rebase exposed a real collision that needs careful, compile-validated reconstruction (not a blind re-push):

**What happened.** Main refactored `mixedbread::SearchOptions` (added `score_threshold` / `return_metadata`, kept `rerank: bool`). This PR replaces `rerank: bool` with a new `Rerank` enum (`Toggle(bool) | Model { model }`, untagged so the wire body stays `bool | object`) plus `DEFAULT_RERANK_MODEL`. During the rebase the `mixedbread/src/lib.rs` (and `polars-mixedbread/src/lib.rs`) conflict got resolved to main's side, **dropping the `Rerank` type entirely**, so `search-core/src/lib.rs` fails to build: `error[E0432]: unresolved imports mixedbread::DEFAULT_RERANK_MODEL, mixedbread::Rerank`.

**What finishing it needs** (the API-trailing-param fix from the AI review is already applied and reviewed-clean):
1. Re-add `Rerank` + `DEFAULT_RERANK_MODEL` + impl to `mixedbread/src/lib.rs` (verbatim from commit 53cf31c4), change `SearchOptions.rerank: bool` → `Rerank`, drop `Copy` from that derive, and **keep** main's `score_threshold`/`return_metadata` fields.
2. Re-verify the auto-merged `search-core/{adapter,backend,search}.rs`, `polars-mixedbread/src/lib.rs`, and the 5 `SearchOptions { ... }` construction sites set both `rerank: Rerank` and main's new fields.
3. Validate on x86_64-linux (the shipped `rerank_serializes_as_bool_or_object` test guards the wire format) — local darwin can't build the `.#checks.x86_64-linux.*` units.

I didn't rush this blind because it's wire-format-sensitive and main is churning, which makes CI-only iteration unreliable. Original pre-rebase commit for reference: 39c305a2.

(triaged via Claude / Opus 4.8)

--- a/packages/mixedbread/src/lib.rs
+++ b/packages/mixedbread/src/lib.rs
@@ -122,14 +122,66 @@ pub enum Error {
 /// Result alias defaulting to this crate's [`Error`].
 pub type Result<T, E = Error> = std::result::Result<T, E>;
 
+/// Default second-stage reranking model: Mixedbread's listwise reranker.
+///
+/// It reads the candidate set as a whole and lifts ranking quality across every
+/// benchmark relative to the prior pointwise default.
+/// See <https://www.mixedbread.com/blog/mxbai-rerank-v3-listwise>.
+pub const DEFAULT_RERANK_MODEL: &str = "mixedbread-ai/mxbai-rerank-v3-listwise";
+
+/// Second-stage reranker selection.
+///
+/// Serialized as the API's `rerank` field, which is `boolean | object`:
+/// [`Rerank::Toggle`] serializes to a bare bool (so the legacy "just turn it
+/// on/off" wire body is byte-for-byte unchanged), while [`Rerank::Model`]
+/// serializes to `{ "model": "..." }` to pin a specific reranking model.
+#[derive(Debug, Clone, serde::Serialize)]
+#[serde(untagged)]
+pub enum Rerank {
+    /// Toggle the server's default reranker: `false` disables reranking,
+    /// `true` applies the API-chosen default model.
+    Toggle(bool),
+    /// Apply a named reranking model, e.g. [`DEFAULT_RERANK_MODEL`].
+    Model {
+        /// Reranking model name forwarded to the API.
+        model: String,
+    },
+}
+
+impl Rerank {
+    /// Reranking disabled (wire `false`).
+    #[must_use]
+    pub const fn off() -> Self {
+        Self::Toggle(false)
+    }
+
+    /// The API's default reranker (wire `true`).
+    #[must_use]
+    pub const fn server_default() -> Self {
+        Self::Toggle(true)
+    }
+
+    /// Pin a specific reranking model.
+    #[must_use]
+    pub fn model(name: impl Into<String>) -> Self {
+        Self::Model { model: name.into() }
+    }
+
+    /// The listwise reranker ([`DEFAULT_RERANK_MODEL`]).
+    #[must_use]
+    pub fn listwise() -> Self {
+        Self::model(DEFAULT_RERANK_MODEL)
+    }
+}
+
 /// Search tuning forwarded to the API.
 ///
 /// `score_threshold` and `return_metadata` are skipped when unset, so a caller
 /// that only sets `rerank`/`agentic` produces the same wire body as before.
-#[derive(Debug, Clone, Copy, serde::Serialize)]
+#[derive(Debug, Clone, serde::Serialize)]
 pub struct SearchOptions {
-    /// Apply the second-stage reranker.
-    pub rerank: bool,
+    /// Apply the second-stage reranker (toggle or a pinned model).
+    pub rerank: Rerank,
     /// Let the API plan and run multiple searches.
     pub agentic: bool,
     /// Drop hits scoring below this threshold (`0.0..=1.0`). Used to keep a
@@ -754,7 +806,28 @@ mod tests {
     use axum::http::{StatusCode, header};
     use axum::response::IntoResponse;
 
-    use super::{BACKOFF_BASE, BACKOFF_CAP, Chunk, Client, Error, MAX_RETRIES, RawChunk, backoff};
+    use super::{
+        BACKOFF_BASE, BACKOFF_CAP, Chunk, Client, DEFAULT_RERANK_MODEL, Error, MAX_RETRIES,
+        RawChunk, Rerank, backoff,
+    };
+
+    #[test]
+    fn rerank_serializes_as_bool_or_object() {
+        // The toggle keeps the legacy bare-bool wire body byte-for-byte.
+        assert_eq!(
+            serde_json::to_value(Rerank::off()).expect("serialize"),
+            serde_json::json!(false)
+        );
+        assert_eq!(
+            serde_json::to_value(Rerank::server_default()).expect("serialize"),
+            serde_json::json!(true)
+        );
+        // A pinned model serializes to the `{ "model": ... }` object form.
+        assert_eq!(
+            serde_json::to_value(Rerank::listwise()).expect("serialize"),
+            serde_json::json!({ "model": DEFAULT_RERANK_MODEL })
+        );
+    }
 
     /// A spawned test server: its base URL and a shared counter of requests it received.
     struct MockServer {

--- a/packages/polars-mixedbread/python/polars_mixedbread/_polars_mixedbread.pyi
+++ b/packages/polars-mixedbread/python/polars_mixedbread/_polars_mixedbread.pyi
@@ -10,10 +10,10 @@ def search_mixedbread(
     top_k: int = ...,
     base_url: str | None = ...,
     rerank: bool = ...,
-    reranker: str | None = ...,
     agentic: bool = ...,
     score_threshold: float | None = ...,
     filters: str | None = ...,
+    reranker: str | None = ...,
 ) -> dict[str, list[Any]]:
     """Run a Mixedbread store search, returning a dict of the six source columns.
 

--- a/packages/polars-mixedbread/src/lib.rs
+++ b/packages/polars-mixedbread/src/lib.rs
@@ -14,7 +14,7 @@
 //! no `pyo3-polars` Arrow-FFI lockstep to keep in sync (unlike `polars-sftp`,
 //! which decodes files in Rust and so must pin both).
 
-use mixedbread::{Client, SearchOptions, filter::Filter};
+use mixedbread::{Client, Rerank, SearchOptions, filter::Filter};
 use pyo3::exceptions::{PyRuntimeError, PyValueError};
 use pyo3::prelude::*;
 use pyo3::types::{PyDict, PyList};
@@ -42,6 +42,9 @@ use pyo3::types::{PyDict, PyList};
     agentic = false,
     score_threshold = None,
     filters = None,
+    // Trailing optional so existing positional callers keep their slots; a name
+    // pins a reranking model, unset uses the listwise default.
+    reranker = None,
 ))]
 #[allow(
     clippy::too_many_arguments,
@@ -58,6 +61,7 @@ fn search_mixedbread(
     agentic: bool,
     score_threshold: Option<f32>,
     filters: Option<String>,
+    reranker: Option<String>,
 ) -> PyResult<Py<PyDict>> {
     if stores.is_empty() {
         return Err(PyValueError::new_err(
@@ -73,6 +77,13 @@ fn search_mixedbread(
         .map_err(|error| {
             PyValueError::new_err(format!("polars-mixedbread: invalid filters JSON: {error}"))
         })?;
+    // `rerank=False` disables reranking; otherwise a named model wins, defaulting
+    // to the listwise reranker for the best ordering.
+    let rerank = match (rerank, reranker) {
+        (false, _) => Rerank::off(),
+        (true, Some(model)) => Rerank::model(model),
+        (true, None) => Rerank::listwise(),
+    };
     let options = SearchOptions {
         rerank,
         agentic,

--- a/packages/search/src/main.rs
+++ b/packages/search/src/main.rs
@@ -14,8 +14,8 @@ use clap::error::ErrorKind;
 use clap::{Args, CommandFactory, Parser, Subcommand};
 use indicatif::ProgressBar;
 use search_core::{
-    CodeScope, DEFAULT_STORE, DisplayHit, Filter, FilterSpec, GrepOptions, GrepTargets, Manifest,
-    MixedbreadStore, SearchOptions, Source, build_filter,
+    CodeScope, DEFAULT_RERANK_MODEL, DEFAULT_STORE, DisplayHit, Filter, FilterSpec, GrepOptions,
+    GrepTargets, Manifest, MixedbreadStore, Rerank, SearchOptions, Source, build_filter,
 };
 
 /// Command-line arguments.
@@ -162,6 +162,11 @@ struct SemanticArgs {
     #[arg(long = "no-rerank")]
     no_rerank: bool,
 
+    /// Reranking model to apply. Defaults to Mixedbread's listwise reranker.
+    /// Ignored when `--no-rerank` is set.
+    #[arg(long = "reranker", default_value_t = DEFAULT_RERANK_MODEL.to_owned())]
+    reranker: String,
+
     /// Include web results from the hosted web store.
     #[arg(short = 'w', long)]
     web: bool,
@@ -281,7 +286,11 @@ async fn run(cli: SemanticArgs) -> anyhow::Result<()> {
     // the manifest is empty (it only ever held this checkout's hashes).
     let manifest = Manifest::default();
     let options = SearchOptions {
-        rerank: !cli.no_rerank,
+        rerank: if cli.no_rerank {
+            Rerank::off()
+        } else {
+            Rerank::model(cli.reranker)
+        },
         agentic: cli.agentic,
     };
     let top_k = cli.max_count.max(1);

